<a href="https://colab.research.google.com/github/forexms78/AI-05-/blob/main/%EB%AF%B8%EC%85%984_2%ED%8C%80_%EB%B0%95%EB%B3%91%ED%98%B8.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
from urllib.request import urlretrieve; urlretrieve("https://raw.githubusercontent.com/c0z0c/jupyter_hangul/refs/heads/beta/helper_c0z0c_dev.py", "helper_c0z0c_dev.py")
import helper_c0z0c_dev as helper

🌐 https://c0z0c.github.io/jupyter_hangul
ℹ️ NumPy 2.0.2 (v2.x+): 호환성 모드 적용됨
install fonts-nanum...
Mounted at /content/drive
✅ 설정 완료: 한글 폰트, plt 전역 등록, pandas 확장, 캐시 기능
pd commit 저장 경로 = /content/drive/MyDrive


포트루칼 은행이 전화 텔레마케팅을 통해 정기 예금 상품을 홍보한 결과 어떠한 고객이 정기 예금에 가입할 가능성이 높은지 컬럼에 있는 고객 정보와 마케팅 정보를 활용하여 y컬럼을 예측하는 모델을 만들기

| 컬럼명         | 설명                                  |
|----------------|---------------------------------------|
| age            | 나이 (숫자)                           |
| job            | 직업 (범주형)                         |
| marital        | 결혼 여부 (범주형)                    |
| education      | 교육 수준 (범주형)                    |
| default        | 신용 불량 여부 (범주형)                |
| housing        | 주택 대출 여부 (범주형)                |
| loan           | 개인 대출 여부 (범주형)                |
| contact        | 연락 유형 (범주형)                     |
| month          | 마지막 연락 월 (범주형)                |
| day_of_week    | 마지막 연락 요일 (범주형)              |
| duration       | 마지막 연락 지속 시간, 초 단위 (숫자)  |
| campaign       | 캠페인 동안 연락 횟수 (숫자)           |
| pdays          | 이전 캠페인 후 지난 일수 (숫자)        |
| previous       | 이전 캠페인 동안 연락 횟수 (숫자)      |
| poutcome       | 이전 캠페인의 결과 (범주형)            |
| emp.var.rate   | 고용 변동률 (숫자)                     |
| cons.price.idx | 소비자 물가지수 (숫자)                 |
| cons.conf.idx  | 소비자 신뢰지수 (숫자)                 |
| euribor3m      | 3개월 유리보 금리 (숫자)              |
| nr.employed    | 고용자 수 (숫자)                       |
| y              | 정기 예금 가입 여부 ('yes' 또는 'no') |

컬럼은 크게 3가지 분류로 나눌 수 있습니다 고객 정보, 마케팅 정보, 사회 및 경제 환경
고객 정보 컬럼은 나이, 직업, 결혼상태, 교육수준, 신용불량여부, 주택대출여부, 개인대출여부
마케팅 정보로는 연락유형, 마지막 연락 월, 요일, 통화시간 이번캠패엔에서 연락한 횟수 이전캠페인에서 연락한 회수, 경과횟수, 결과
사회 및 경제 환경으론 고용변동률, 소비자 물가지수, 소비자 신뢰지수 3개월 유리보금리, 고용자수
타겟변수는 y정기예금가입여부

In [2]:
# 데이터 불러오기

from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [3]:
# 분류알고리즘을 사용하면된다 KNN 로지스틱회귀 xboost

import pandas as pd
df = pd.read_csv('/content/drive/MyDrive/ai05/bank-additional-full.csv',sep=';')

In [4]:
df.head(20)

age          job   marital            education  default housing loan  \
0    56    housemaid   married             basic.4y       no      no   no   
1    57     services   married          high.school  unknown      no   no   
2    37     services   married          high.school       no     yes   no   
3    40       admin.   married             basic.6y       no      no   no   
4    56     services   married          high.school       no      no  yes   
5    45     services   married             basic.9y  unknown      no   no   
6    59       admin.   married  professional.course       no      no   no   
7    41  blue-collar   married              unknown  unknown      no   no   
8    24   technician    single  professional.course       no     yes   no   
9    25     services    single          high.school       no     yes   no   
10   41  blue-collar   married              unknown  unknown      no   no   
11   25     services    single          high.school       no     yes   no   
12   29  blue-collar    single          high.school       no      no  yes   
13   57    housemaid  divorced             basic.4y       no     yes   no   
14   35  blue-collar   married             basic.6y       no     yes   no   
15   54      retired   married             basic.9y  unknown     yes  yes   
16   35  blue-collar   married             basic.6y       no     yes   no   
17   46  blue-collar   married             basic.6y  unknown     yes  yes   
18   50  blue-collar   married             basic.9y       no     yes  yes   
19   39   management    single             basic.9y  unknown      no   no   

      contact month day_of_week  duration  campaign  pdays  previous  \
0   telephone   may         mon       261         1    999         0   
1   telephone   may         mon       149         1    999         0   
2   telephone   may         mon       226         1    999         0   
3   telephone   may         mon       151         1    999         0   
4   telephone   may         mon       307         1    999         0   
5   telephone   may         mon       198         1    999         0   
6   telephone   may         mon       139         1    999         0   
7   telephone   may         mon       217         1    999         0   
8   telephone   may         mon       380         1    999         0   
9   telephone   may         mon        50         1    999         0   
10  telephone   may         mon        55         1    999         0   
11  telephone   may         mon       222         1    999         0   
12  telephone   may         mon       137         1    999         0   
13  telephone   may         mon       293         1    999         0   
14  telephone   may         mon       146         1    999         0   
15  telephone   may         mon       174         1    999         0   
16  telephone   may         mon       312         1    999         0   
17  telephone   may         mon       440         1    999         0   
18  telephone   may         mon       353         1    999         0   
19  telephone   may         mon       195         1    999         0   

       poutcome  emp.var.rate  cons.price.idx  cons.conf.idx  euribor3m  \
0   nonexistent           1.1          93.994          -36.4      4.857   
1   nonexistent           1.1          93.994          -36.4      4.857   
2   nonexistent           1.1          93.994          -36.4      4.857   
3   nonexistent           1.1          93.994          -36.4      4.857   
4   nonexistent           1.1          93.994          -36.4      4.857   
5   nonexistent           1.1          93.994          -36.4      4.857   
6   nonexistent           1.1          93.994          -36.4      4.857   
7   nonexistent           1.1          93.994          -36.4      4.857   
8   nonexistent           1.1          93.994          -36.4      4.857   
9   nonexistent           1.1          93.994          -36.4      4.857   
10  nonexistent           1.1          93.994   

In [5]:
df.describe()

age      duration      campaign         pdays      previous  \
count  41188.00000  41188.000000  41188.000000  41188.000000  41188.000000   
mean      40.02406    258.285010      2.567593    962.475454      0.172963   
std       10.42125    259.279249      2.770014    186.910907      0.494901   
min       17.00000      0.000000      1.000000      0.000000      0.000000   
25%       32.00000    102.000000      1.000000    999.000000      0.000000   
50%       38.00000    180.000000      2.000000    999.000000      0.000000   
75%       47.00000    319.000000      3.000000    999.000000      0.000000   
max       98.00000   4918.000000     56.000000    999.000000      7.000000   

       emp.var.rate  cons.price.idx  cons.conf.idx     euribor3m   nr.employed  
count  41188.000000    41188.000000   41188.000000  41188.000000  41188.000000  
mean       0.081886       93.575664     -40.502600      3.621291   5167.035911  
std        1.570960        0.578840       4.628198      1.734447     72.251528  
min       -3.400000       92.201000     -50.800000      0.634000   4963.600000  
25%       -1.800000       93.075000     -42.700000      1.344000   5099.100000  
50%        1.100000       93.749000     -41.800000      4.857000   5191.000000  
75%        1.400000       93.994000     -36.400000      4.961000   5228.100000  
max        1.400000       94.767000     -26.900000      5.045000   5228.100000

age에 대해선 크게 중요한걸 못느끼는게 최소나의와 최대나이가 크게 벗어나지 않아서 넘어가려고 합니다

duration 에선 통화시간이 0 초인것은 연결되지 않았다고 판단해서 예금을 가입했을 확률이 없다고 판단합니다
4918초는 상당히 오래 통화한것으로 보아 예금에 상당히 많은 관심을 가지고 있다고 판단할 수 있습니다 duration 컬럼은 학습에는 필요하지 않습니다
이유는 통화하기 전에 예금을 가입할지 예측하는것이 때문입니다 duration은 EDA측면에서는 매우 유용하지만 학습에서는 제외합니다

campaign같은 경우는 연락한 횟수입니다 전화를 많이걸었다는뜻은 전화를 받지않아서 여러번 걸었다는 뜻으로 해석할 수 있습니다

pdays 이전캠페인 후 현재 캠페인까지의 경화한 일수 pday같은 경우는 999는 새로운 고객이라고 볼 수 있습니다 시간이 경과된걸 알 수 없기때문입니다
999값을 를 처리할때 2가지로 볼 수있습니다
1. boolean 형태로 0과 1을 사용하여 접촉을 했느냐 접촉하지 않았느냐
2. 999데이터를 제외한값을 log화 해서 사용하느냐

previous는 이번캠페인 이전에 고객에게 연락했던 횟수입니다
예를들어 previous 가 0이면  pdays컬럼값은 999가 되어야합니다 전처리 진행시 이값에 위배되는 값이 있는지 파악해야합니다 이컬럼이 1이상이면 과거 은행에 캠페인에 참여한 이력이 있다는 말입니다 pday랑 다르게 횟수인 점을 확인해야합니다

emp.var.rate는 고용 변동률을 나타냅니다 양수면 고용이 증가하고 음수면 고용의 감소를 나타냅니다 고용률이 높을때 투자자들이 예금 가입할 확률에 대해서 알아볼 필요성과 그 반대의 상황도 확인해보기 좋을거 같습니다

price.idx는 물가 지수를 나타내며 기준은 100을 기준으로합니다 92이라는 숫자는 물가지수가 92%나타내는거빈다

conf.idx는 소비자 신뢰지수로써 소비자가 -40% 만큼 경제를 부정적으로 본다는 것을 의미합니다

eruibor3m 금리를 나타냅니다







In [6]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 41188 entries, 0 to 41187
Data columns (total 21 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   age             41188 non-null  int64  
 1   job             41188 non-null  object 
 2   marital         41188 non-null  object 
 3   education       41188 non-null  object 
 4   default         41188 non-null  object 
 5   housing         41188 non-null  object 
 6   loan            41188 non-null  object 
 7   contact         41188 non-null  object 
 8   month           41188 non-null  object 
 9   day_of_week     41188 non-null  object 
 10  duration        41188 non-null  int64  
 11  campaign        41188 non-null  int64  
 12  pdays           41188 non-null  int64  
 13  previous        41188 non-null  int64  
 14  poutcome        41188 non-null  object 
 15  emp.var.rate    41188 non-null  float64
 16  cons.price.idx  41188 non-null  float64
 17  cons.conf.idx   41188 non-null 

# 고객특성 파악

# 마케팅 전략

# 핵심 변수 파악